source/lambda/apiprocessor/redaction.py

```diff
@@ -41,40 +57,242 @@ def getDocumentRedaction(documentId, bucket):
+
+    if isinstance(item['page'], int) == False:
+        return (400, "bad request, redacted item has invalid page number type")
+
+    if item['page'] < 1:
+        return (400, "bad request, redacted item page number is invalid")
+    
+    # validate top number
+    if 'top' not in item:
+        return (400, "bad request, redacted item has no top number")
+
+    if isinstance(item['top'], float) == False:
+        return (400, "bad request, redacted item has invalid top number type")
+
+    if item['top'] < 0.0:
+        return (400, "bad request, redacted item top number is invalid")
+
+    # validate left number
+    if 'left' not in item:
+        return (400, "bad request, redacted item has no left number")
+
+    if isinstance(item['left'], float) == False:
+        return (400, "bad request, redacted item has invalid left number type")
+
+    if item['left'] < 0.0:
+        return (400, "bad request, redacted item left number is invalid")
+            
+    # validate width number
+    if 'width' not in item:
+        return (400, "bad request, redacted item has no width number")
+
+    if isinstance(item['width'], float) == False:
+        return (400, "bad request, redacted item has invalid width number type")
+
+    if item['width'] < 0.0:
+        return (400, "bad request, redacted item width number is invalid")
+
+    # validate height number
+    if 'height' not in item:
+        return (400, "bad request, redacted item has no height number")
+
+    if isinstance(item['height'], float) == False:
+        return (400, "bad request, redacted item has invalid height number type")
+
+    if item['height'] < 0.0:
+        return (400, "bad request, redacted item height number is invalid")
+
+   # validate label string
+    if 'label' not in item:
+        return (400, "bad request, redacted item has no label")
+
+    if isinstance(item['label'], str) == False:
+        return (400, "bad request, redacted item label is not a string")
+
+    return (200, "ok")
+
+
 # retrieve the redacted items of this document
-def saveDocumentRedaction(documentId, bucket, body):
+def saveDocumentRedaction(documentId, bucket, table, body):
 
-    result = {}
-    result['status'] = False
+    statusCode = 400
+    result = "bad request"
+    
+    #pdb.set_trace()
     
-    # save the document redaction, if it parses correctly
+    # save the document redaction, if it parses and validates correctly
     try:
         redaction = json.loads(body)
         
-        S3Helper.writeToS3(json.dumps(redaction),
-                           bucket,
-                           'public/' + documentId + '/redaction.json')
-    
-        result['status'] = True
-    
     except Exception as e:
-        print("saveDocumentRedaction exception: " + str(e))
+        return (400, "invalid json document")
+        
+    # validate redaction
+    if len(redaction.keys()) != REDACTION_NUM_OF_KEYS:
+        return (400, "bad request, wrong number of redaction keys")
```

Comment:
Should the message also state what is the "right" or expected number of redaction keys?

Reply:
Also rather than the numbers, can we just validate the exact keys of Redaction that are required?

Reply:
ok, I think it's better to move that check at the very end, after all the valid entries have been validated.  Then if there are extra keys, we say that there is extra data that is not supported, to only provide uuid, headers, footers and redactedItems

Reply:
I would like to take the fail fast approach in case of an error.
If we first validate if we have all the required keys, and then do specific key validation, we can catch on errors quicker taking a top down approach.

Thoughts?

Reply:
ok, I will move it back to where it was, with the added information for logging